umbra-js/src/classes/Umbra.ts

```diff
@@ -175,55 +154,123 @@ export class Umbra {
+        .sendToken(
+          stealthKeyPair.address,
+          token,
+          amount,
+          compressedXCoordinate,
+          encrypted.ciphertext,
+          txOverrides
+        );
     }
 
     // We do not wait for the transaction to be mined before returning it
     return { tx, stealthKeyPair };
   }
 
   /**
-   * @notice Withdraw funds for a recipient
+   * @notice Withdraw ETH or tokens to a specified destination address with a regular transaction
+   * @param generationPrivateKey Receiver's generation private key
    * @param token Address of token to withdraw
-   * @param recipient Recipient address
+   * @param stealthAddress Stealth address funds were sent to
+   * @param destination Address to send funds to
    * @param overrides Override the payload extension, gas limit, gas price, or nonce
+   * @dev The provider used for sending the transaction will be the one associated with the Umbra
+   * instance. Use the `connect()` method to explicitly set the provider you want to use before
+   * calling this method, e.g. `umbra.connect(myProvider).withdraw(...)
+   * @dev This method does not relay meta-transactions
    */
-  withdraw(token: string, recipient: string, overrides: UmbraOverrides = {}) {
-    console.log(token, recipient, overrides);
+  async withdraw(
+    generationPrivateKey: string,
+    token: string,
+    stealthAddress: string,
+    destination: string,
+    overrides: Overrides = {}
+  ) {
+    // Configure signer
+    const wallet = new Wallet(generationPrivateKey);
+    const txSigner = this.getSigner(wallet);
 
-    // Determine if recipient caller is withdrawing token or ETH by reading contract
-    // TODO
+    if (isETH(token)) {
+      // Withdraw ETH
+      // Based on gas price, compute how much ETH to transfer to avoid dust
+      const ethBalance = await this.provider.getBalance(stealthAddress);
+      const gasPrice = BigNumber.from(overrides.gasPrice || (await this.provider.getGasPrice()));
+      const gasLimit = BigNumber.from(overrides.gasLimit || '21000');
+      const txCost = gasPrice.mul(gasLimit);
+      return await txSigner.sendTransaction({
+        to: destination,
+        value: ethBalance.sub(txCost),
+        gasPrice,
+        gasLimit,
+        nonce: overrides.nonce || undefined, // nonce will be determined by ethers if undefined
+      });
+    } else {
+      // Withdrawing a token
+      return await this.umbraContract.connect(txSigner).withdrawToken(destination, overrides);
+    }
+  }
 
-    // // Withdraw funds
-    // if (isETH(token)) {
-    //   // Based on gas price, compute how much ETH to transfer to avoid dust
-    //   // TODO
-    // } else {
-    //   // this.umbra.withdrawToken(...);
-    // }
+  /**
+   * @notice Withdraw tokens by sending a meta-transaction on behalf of a user
+   * @param signer Signer to send transaction from
+   * @param stealthAddr Stealth address funds were sent to
+   * @param destination Address to send funds to
```

Comment:
Same comment as above.

Reply:
Done in ba0e6cbc548a145aa90dc54534b6baef8f192ff5